### DISCUSSION TITLE
when is typebox expected to hit 1.0 (aka. be stable)?

### DISCUSSION THREAD

**Discussion by boywithkeyboard** [sinclairzx81/typebox#298] [Q&A] [answered]
what's missing for a stable release? i'd love to help you work on a 1.0 release :)

**Comment by andreykrupskii** [comment]
I'm also curious about it 😉 

Our company has policies that prevent us from using 0.x libraries. So, I hope a stable release is going to be published soon 😊

**Comment by boywithkeyboard** [comment, marked as answer]
@andreykrupskii this can probably be seen as an answer on this discussion imo: https://github.com/sinclairzx81/typebox/issues/299#issuecomment-1368729238

**Reply by sinclairzx81** [reply]
@unvented @andreykrupskii Hi! Apologies for the delay in reply (I've been on holiday)

### Outstanding

So to bring some visibility on the 1.0 release, these are the outstanding items that would constitute a 1.0:

- Restructure Type composition backend to better allow union/intersection types to compose with utility types.
  - Resolution of Intersecting of Union Type issue (requires support for draft 2019-09 minimum)
- Investigate programmable type system for user defined JIT compilation (via some abstract instruction builder, currently being prototyped in private repository)
- Investigate LCS algorithm for efficient array diffing (Value.Diff / Value.Patch, networking)
- Finalize on FNV1A64 hashing algorithm (Value.Hash, used for hash codes, but am looking for faster, relatively non-collision hashing algos)
- Minor project refactoring (mostly removing duplicate type guards)
- A website

These outstanding items have quite a bit of detail and history around them, so happy to discuss them in more depth (maybe via separate discussion thread)

### Schedule

In terms of timings, the 1.0 release is tentatively planned for a 2nd, 3rd quarter this year. However this largely comes down to contractual obligations I have throughout the year (noting TypeBox usually only gets significant updates in between contractual work I do).

### Stability

The library is actually very stable today despite being under 1.0. But do note that TypeBox does enjoy a certain level of flexibility keeping under 1.0 for now which allows it to slowly and strategically to roll out new features with the occasional (documented) breaking change where necessary inline with the vision for the library (see Outstanding). Note, there's a little bit of concern of losing flexibility by preemptively jumping to 1.0 as it would imply the library was entering a general maintenance state, of which I only want to maintaining the envisioned feature set (and after a few months testing the 1.0 feature set in the wild).

### Website

One of the main blockers for a 1.0 release is a website. Historically, it's been very challenging to condense all information in the readme into a digestible form. I think there's enough functionality in TypeBox now to probably start thinking about a website and could certainly use some assistance and guidance putting one together. I really open to community thoughts here.

Hope that sheds a little bit of light on where things are at (Jan, 2023)
Cheers!
S

**Reply by andreykrupskii** [reply]
@sinclairzx81, thank you for such a detailed response 👍 I think that it would be useful to have a document with all planned items for v1, so the community will be able to help and contribute 😊

**Reply by Xhale1** [reply]
@sinclairzx81 I'd be happy to help with a website when you decide it's time :)

**Reply by paulius-valiunas** [reply]
May I suggest opening a tracking issue or adding a milestone to track progress towards the release?
